c/hebcal.1
.\" $Id$  -*- nroff -*-
.\" Copyright(c) 1994-2006  Danny Sadinoff
.\" Portions Copyright (c) 2002 Michael J. Radwin. All Rights Reserved.
.\" See section COPYING for conditions for redistribution
.TH HEBCAL 1 "Hebcal Version 3.5" "Danny Sadinoff"
.\" "
.de BP
.sp
.ti \-.2i
\(**
..
.SH NAME
hebcal - a Jewish calendar generator
.SH SYNOPSIS
.B hebcal
[ 
.B \-8acdDehHiorsStTwy
]
.if n .ti +5
.br
[ 
.B \-I
.BI input_file
]
.if n .ti +5
.br
[ 
.B \-Y
.BI yahrtzeit_file
]
.if n .ti +5
.br
[
.B \-C 
.BI city
]
.if n .ti +5
.br
[
.B \-l 
.BI latitude 
.B \-L 
.BI longitude]
.if n .ti +5
.br
[
.B \-z 
.BI timezone
]
.if n .ti +5
.br
[
.B \-Z
.BI daylight_savings_scheme
]
.if n .ti +5
.br
[[
.BI month
[
.BI day
]]
.BI year
]
.br
.B hebcal help
.br
.B hebcal info
.br
.B hebcal DST
.br
.B hebcal cities
.br
.B hebcal copying
.br
.B hebcal warranty
.PP
.SH DESCRIPTION
With no arguments, 
.B hebcal
will print to stdout the 
dates of the Jewish holidays in the current
secular year.  Each line is prefixed with a gregorian date of the form
mm/dd/yyyy.  
.PP
By specifying 
.BI month,
.BI day,
or 
.BI year,
output can be limited to a particular month or date
in a particular year.  Note that
.BI year
is usually a four-digit integer,  So 92 is during the Roman period,
not the late twentieth century.  In if the hebrew dates option is
turned on, this number represents th Jewish calendar year.
.BI month
is a number from 1..12, or the name of a Jewish calendar month.
.sp
.BI day 
is a number from 1..31.
.sp
For example, the command
.nf
	hebcal 10 1992
.fi
will print out the holidays occurring in October of 1992 C.E., while
the command
.nf

	hebcal Tish 5752

.fi
will print dates of interest in the month of Tishrei in Jewish
calendar year 5752.
.PP
NOTE:  hebcal 92 is 
.I not
the same as hebcal 1992.  The year is assumed to be complete, so the
former calendar preceeds the latter by nineteen centuries.
.PP
A few other bells and whistles include the weekly 
.I sedra
as well as the day of the week, the count of the 
.I omer,
and the Hebrew date.
.PP
Output from hebcal can be used to drive calendar(1).  
Day\-to\-day use for hebcal is provided for in the \c
.B -T \c
and \c
.B -t \c
switches, which print out Jewish calendar entries for the current
date.
.PP
To get a quick\-reference on\-line help, type 
.nf
	hebcal help
.fi
at the command prompt.
.SH OPTIONS
.TP
.B "\-8"
Use 8-bit Hebrew (ISO-8859-8-Logical).
.TP
.BI "\-a"
Use Ashkenazis hebrew.
.TP
.B "\-c"
Add \c
.I approximate \c
candle-lighting times.  See below.
.TP
.B "\-d"
Print the Hebrew date for the entire date range.
.TP
.B "\-D"
Print the hebrew date for dates with some event.
.TP
.B "\-e"
Change the output format to European\-style dates: dd.mm.yyyy
.TP
.B "\-h"
Suppress holidays in output.  User defined calendar events are
unaffected by this switch.
.TP
.B "\-H"
When the \-H switch is used, all dates specified on the command line
are assumed to be hebrew dates.  So for instance,
.sp
.ft B
\	example% hebcal \-H 5754
.sp
will print data for 5754, Starting in Tishrei, and ending in Elul.
Hebcal is smart enough to detect a Hebrew month and infer that you
want a Hebrew date range, so you could type
.sp
.ft B
\	example% hebcal tish 5754
.sp
The \-H switch would be superfluous in this case.  Invoking hebcal with
just the \-H switch by itself will print data for the current Hebrew
year, starting in Tishrei.  
.TP
.B "\-i"
Use the Israeli sedra scheme when used in conjunction with \c
.B \-S \c
or \c
.B \-s.
This has no effect if the \c
.B \-S \c
or \c
.B \-s \c
switches are unused.
.TP
.BI "\-I " "file"
Read extra events from \c
.I "file."
These events are printed regardless of the \c
.B \-h \c
suppress holidays switch.
.br
There is one holiday per line in 
.BI file,
each with the format
.br
.br
.I "month\ day\ description"
.br
.br
where \c
.I month \c
is a string identifying the Jewish month in question \c
.I day \c
is a number from 1 to 30, and \c
.I description
is a newline-terminated string describing the holiday.
An example might be 
.br
.br
\	Adar 1 Start cleaning kitchen for Passover.
.br
\	Adar 1 Start cleaning kitchen for Passover.
.TP
.BI "\-l " "deg,min"
Set the latitude for solar calculations to \c
.I deg \c
degrees and \c
.I min \c minutes.  Negative values are south.
.TP
.BI "\-L " "deg,min"
Set the longitude for solar calculations to \c
.I deg \c
degrees and \c
.I min \c
minutes.  
.B NOTE: Negative values are EAST longitude.
.TP
.B "\-o"
Add the count of the 
.I omer
to the output.
.TP
.B "\-r"
Use a tab-delineated format, and somewhat terser
strings.  Instead of saying \&\|' 13th day of the omer \&\|'
hebcal will say \&\|' Omer: 13 \&\|'
.TP
.B "\-s"
Add the weekly 
.I sedra 
to the output on Saturdays.  See \c
.B \-i.
.TP
.B "\-S"
Add the weekly 
.I sedra 
to the output every day.  When this option is invoked, every time a
day is printed, the torah reading for the Saturday on or immediately
following that date is printed.  If there is no reading for the next
Saturday, then nothing is printed.  See
.B \-i.
.TP
.B "\-t"
Print calendar information for today's date only.  
.B \-d 
and
.B \-o
are asserted with this option. 
.TP
.B "\-T"
Same as -t, only without the gregorian date.  This option is useful in
login scripts, just to see what's happening today in the Jewish calendar.
.TP
.B "\-w"
Add the day of the week to the output.
.TP
.B "\-y"
Print only the last two digits of the year.
.TP
.BI "\-Y " "file"
Read a table of yahrtzeit dates from \c
.I "file."
These events are printed regardless of the \c
.B \-h \c
suppress holidays switch.
.br
.br
There is one death-date per line in \c
.I file,
each with the format
.br
.br
.I "month\ day\ year\ description"
.br
.br
where \c
.I month, day \c
and \c
.I year \c
form the  \c
.B gregorian \c 
date of death.
.I description
is a newline-terminated string to be printed on the \c
.I yahrtzeit.
An example might be 
.br

\	12 29 1957 Menachem Mendel's yahrtzeit.
.br
\	5 15 1930 Benjamin's yahrtzeit.
.SH "CANDLE\-LIGHTING TIMES"
Hebcal's candlelighting times are only approximations.
If you ever have any doubts about it's times, consult
your local halachic authority.  If you enter geographic
coordinates above the artic circle or antarctic circle,
the times are guaranteed to be wrong.
.PP
Hebcal contains a small database of cities with their
associated geographic information and time-zone information.
The geographic and time information necessary to
calculate sundown times can come to hebcal any of three
ways:
.PP
1) The default: the system manager sets a default city
when the program is compiled.
.br
2) Hebcal looks in the environment variable 
.B HEBCAL_CITY
for the name of a city in hebcal's database, and if it
finds one, hebcal will make that the new default city.
.br
3) 1 and 2 may be overridden by command line arguments, including
those specified in the
.B HEBCAL_OPTS
environment variable.
The most natural way to do this is to use the \c
.BI "\-c " "city " \c
command.  This will localize hebcal to \c
.I "city. "
A list of the cities hebcal knows about can be obtained
by typing
.nf
	hebcal cities
.fi
at the command prompt.  If the city you want isn't on that list, you
can directly control hebcal's geographic information with the \c
.B \-l, \-L \-z \c
and \c
.BI "\-Z " "DST " \c
switches.  Note that changing the geographic coordinates
causes the timezone to default to Zulu and the daylight
savings time processor to default to 'none.' To get a
list of possible values for \c
.I DST,
type
.nf
	hebcal DST
.fi
at the command prompt.
.PP
For a status report on customizations, type 
type
.nf
	hebcal info
.fi
at the command prompt.
.SH EXAMPLES
.LP
To find the days of the \c
.I omer \c
in 1997, printing the days of the week:
.RS
.nf
.ft B
example% hebcal -how 1997
.br
4/23/97 Wed, 1st day of the Omer
.br
4/24/97 Thu, 2nd day of the Omer
.br
4/25/97 Fri, 3rd day of the Omer
.br
 .
.br
 .
.br
 .
.br
6/9/97 Mon, 48th day of the Omer
.br
6/10/97 Tue, 49th day of the Omer
.RE
.LP
To print only the weekly \c
.I sedrot \c
of Nisan 5770
.RS
.nf
.ft B
example% hebcal -hs Nisan 5770
.br
3/20/2010 Parashat Vayikra
.br
3/27/2010 Parashat Tzav
.br
4/10/2010 Parashat Shmini
.br
.RE
.LP 
To find out what's happening in the Jewish calendar today , use
.RS
.nf
.ft B
example% hebcal -TS
.br
19 of Nisan, 5752
.br
Parshat Achrei Mot
.br
Pesach V (CH"M)
.\" "
.br
4th day of the Omer
.br
.RE
.SH ENVIRONMENT
.BI Hebcal
uses two environment variables:
.TP 
.B HEBCAL_CITY
.BI "Hebcal " \c
uses this value as the default city for sunset calculations.  A list
of available cities is available with from hebcal with the command:
.br
hebcal cities
.TP
.B HEBCAL_OPTS
The value of this variable is automatically processed as if it were
typed at the command line \c
.I before \c
any other actual command\-line\-arguments.
.SH AUTHOR
Danny Sadinoff
.SH SEE ALSO
calendar(1), emacs(1), hcal(1), hdate(1), omer(1), remind(1), rise(1)
.PP
The latest version of the code will be available from 
http://sourceforge.net/projects/hebcal
.PP
The original motivation for the algorithms in this
program was the \c
.I "Tur Shulchan Aruch.  "
.PP
For version 3, much of the program was rewritten using
Emacs 19's calendar routines by Edward M. Reingold and
Nachum Dershowitz.  Their program is extremely clear and
provides many instructive examples of fine calendar code in
emacs-LISP.
.PP
A well written treatment of the Jewish
calendar for the layman can be found in \c
.I Understanding the Jewish Calendar \c
by Rabbi Nathan Bushwick.  A more complete bibliography on the topic
can be found there, as well as in the \c
.I Encyclopedia Judaica \c
entry on the calendar.  
.SH DIAGNOSTICS
.TP
.B hebcal help
Prints a shorter version of this manpage, with comments on each option.
.TP
.B hebcal info
Prints the version number and default values of the program. 
.TP
.B hebcal DST
Prints a list of available daylight savings time schemes, suitable as
arguments to the \c
.BI "\-Z " "DST " \c
option.
.TP
.B hebcal cities
Prints a list of cities which hebcal knows about, suitable as
arguments to the \c
.BI "\-C " "city " \c
option.  If your city does not appear on this list, put the necessary
defaults in the
.B DST_OPTS \c
 environment variable.
.TP
.B hebcal copying
Prints the GNU license, with information about copying the program.
See below.
.TP
.B hebcal warranty
Tells you how there's NO WARRANTY for hebcal.
.SH DISCLAIMER
This is just a program I wrote during summer school and
while avoiding my senior project.  It should not be invested
with any sort of halachic authority.
.SH BUGS
Hebrew dates are only valid before sundown on that
secular date.  An option to control this will be added
in a later release.
.PP
Negative longitudes are EAST of Greenwich.
.PP
Some combinations of options produce weird results, e.g. 
.br
hebcal -dH nisan 5744
.br
hebcal -dH 5744
.br
This comes into play when you use the 
.B ENV_OPT
environment variable.
.PP
The sunup/sundown routines aren't accurate enough.  If you enter
geographic coordinates above the artic circle or antarctic circle, the
times are guaranteed to be wrong.
.PP
Hebcal only translates between the Gregorian calendar
and the Jewish calendar. This means that the results
will be at least partly useless where and when the
Gregorian calendar was not used, e.g. before the
1752 in Britain and before circa 1918 in Russia.  
See the wikipedia entry for "Daylight saving time" for
a splendid chart depicting when the changeover from 
the Julian to the Gregorian calendars occurred in various places.
.PP
Hebcal cannot handle date computations before 2 C.E.  sorry.
.PP
Daylight-Savings time rules are as up-to-date as a nonpaying job
allows.  US DST rules are correct only back to 1966.  
.PP
Hebcal assumes that the Energy Policy Act of 2005, which changes the DST 
rules in the US will go into effect, even though congress may still revert it.
.SH BUG REPORTS TO
Danny Sadinoff
.br
danny@sadinoff.com
.SH COPYING
Copyright (c) 1994-2006 Danny Sadinoff
.br
Portions Copyright (c) 2002 Michael J. Radwin. All Rights Reserved.
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the Free Software Foundation instead of in
the original English.
.PP
For a full text of the copyright and lack of warranty information, 
type 
.br
hebcal copying
.br
or 
.br
hebcal warranty
.br 
at the command line.


c/hebcal.c
/*
   $Id$
   Hebcal - A Jewish Calendar Generator
   Copyright (C) 1994-2004  Danny Sadinoff
   Portions Copyright (c) 2002 Michael J. Radwin. All Rights Reserved.

   http://sourceforge.net/projects/hebcal

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

   Danny Sadinoff can be reached at 
   danny@sadinoff.com
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "hebcal.h"
#include "common.h"
#include "danlib.h"
#include "myerror.h"
#include "rise.h"
#include "sedra.h"
#include "format.h"

#define LANGUAGE(str) (ashkenazis_sw && (str)[1] ? ((str)[1]) : ((str)[0]))

#define NM_LEN 60

int
  ashkenazis_sw, candleLighting_sw, euroDates_sw, hebrewDates_sw, inputFile_sw,
  israel_sw, latlong_sw, printOmer_sw, sedraAllWeek_sw, sedrot_sw, noGreg_sw,
  printHebDates_sw, printSomeHebDates_sw, noHolidays_sw, tabs_sw, weekday_sw,  suppress_rosh_chodesh_sw,
  yearDigits_sw, yahrtzeitFile_sw, DST_scheme, DST_value;
int iso8859_8_sw;
long beginOmer, endOmer;
FILE *inFile, *yFile;
char *formatString;
extern holstorep_t holidays[14][MAXDAYS], union_Adar[MAXDAYS];

#define SIMCHAT_DAY 23
int havdalah_minutes = 72;      /* default */
int light_offset = -18;         /* outside jerusalem */

/*-------------------------------------------------------------------------*/

year_t yearData( int hebyr )
{
    date_t tempDate;
    year_t retYear;
    
    tempDate.yy = hebyr;
    tempDate.mm = TISHREI;
    tempDate.dd = 1;
    
    retYear.first_day_of_week = (int) (hebrew2abs (tempDate) % 7L);
    retYear.leap_p = LEAP_YR_HEB (hebyr);
    return retYear;
}

/*-------------------------------------------------------------------------*/

date_t nextHebDate( date_t dth )
{
    dth.dd++;
    if (dth.dd > max_days_in_heb_month (dth.mm, dth.yy))
    {
        if (dth.mm == MONTHS_IN_HEB (dth.yy))
        {
            dth.mm = NISAN;
            dth.dd = 1;
        }
        else
        {
            dth.mm++;
            dth.dd = 1;
            if (dth.mm == TISHREI)
            {
                dth.yy++;
            }
        }
    }
    return dth;
}

/*-------------------------------------------------------------------------*/

/* end-of-loop increment routine */
void incHebGregDate( 
    date_t *dtHeb,
    date_t *dtGreg, 
    long *dtAbs, 
    int *wkday, 
    year_t *theYear
    )
{
  incDate (dtGreg, 1L);
  *dtHeb = nextHebDate (*dtHeb);
  (*dtAbs)++;
  *wkday = (*wkday + 1) % 7;

  if (TISHREI == dtHeb->mm &&   /* if Rosh Hashana then reset YearData */
      1 == dtHeb->dd)
    {
      *theYear = yearData (dtHeb->yy);
      if (sedraAllWeek_sw || sedrot_sw)
        reset_sedra (dtHeb->yy);
      if (printOmer_sw)
        reset_Omer (dtHeb->yy);
      init_holidays(dtHeb->yy);
    }
}

/*-------------------------------------------------------------------------*/

void PrintGregDate( date_t dt )
{
    if (!noGreg_sw)
    {
        if (euroDates_sw)
            printf ("%d.%d.", dt.dd, dt.mm);        /* dd/mm/yyyy */
        else
            printf ("%d/%d/", dt.mm, dt.dd);        /* mm/dd/yyyy */
        
        if (yearDigits_sw)
            printf ("%d", dt.yy % 100);
        else
            printf ("%d", dt.yy);
        
        if (tabs_sw)
            putchar ('\t');
        else
            putchar (' ');
    }
    
    if (weekday_sw)
        printf ("%s, ", ShortDayNames[dayOfWeek (dt)]);
}

/*-------------------------------------------------------------------------*/

void set_DST_bounds (long *beginDST, long *endDST, int gregYr)
{
    date_t tempDt;
    
    tempDt.yy = gregYr;
    tempDt.dd = 1;
    tempDt.mm = 4;
    
    switch (DST_scheme)
    {
    case DST_USOFA:
      /* FIX: this needs more work for historical DST, obviously */
      if( gregYr >= 1966 && gregYr < 1987 ) 
	{
        /*  the last sunday in April */
	  tempDt.dd = 30;
	  tempDt.mm = 4;
	  *beginDST = day_on_or_before (SUN, greg2abs (tempDt) );

        /*  the sunday before the first of november */
        tempDt.dd = 1;
        tempDt.mm = 11;
        *endDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);
        DST_value = 60;
        break;
	}
      else if( gregYr >= 1987 && gregYr < 2007 )
	{
        /*  the first sunday in April */
        *beginDST = day_on_or_before (SUN, greg2abs (tempDt) + 6L);


        /*  the sunday before the first of november */
        tempDt.dd = 1;
        tempDt.mm = 11;
        *endDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);

        DST_value = 60;
        break;
	}
      else if( gregYr >= 2007 )
	{
        /*  Second Sunday in March */
        tempDt.dd = 1;
        tempDt.mm = 3;
        *beginDST = day_on_or_before (SUN, greg2abs(tempDt) + 13L);


        /*  first sunday of november */
        tempDt.dd = 1;
        tempDt.mm = 11;
        *endDST = day_on_or_before (SUN, greg2abs(tempDt) +6L);

        DST_value = 60;
        break;
	}
    case DST_EU:
        /* The EU version of Daylight Saving Time runs from the last
         * Sunday in March through the last Sunday in October.
       * it starts at 1am instead of 2am, but that's not relevant here.
       * http://webexhibits.org/daylightsaving/g.html
       */
        *beginDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);
        tempDt.dd = 1;
        tempDt.mm = 11;
        /*  the sunday before the first of november */
        *endDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);
        DST_value = 60;
        break;
    case DST_AUNZ:
        /* Australasia is the inverse of European Union */
        *endDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);
        tempDt.dd = 1;
        tempDt.mm = 11;
        *beginDST = day_on_or_before (SUN, greg2abs (tempDt) - 1L);
        DST_value = 60;
        break;
    case DST_ISRAEL:
        /* daylight savings time begins in Israel on the */
      /* first Sunday after Passover ends on Nisan 21: */
        tempDt.mm = NISAN;
        tempDt.dd = 28;
        tempDt.yy = gregYr + 3760;
        *beginDST = day_on_or_before (SUN, hebrew2abs (tempDt));
        
        /* daylight savings time ends in Israel on the Sunday Selichot */
        /* begins: */
        tempDt.mm = TISHREI;
        tempDt.dd = 1;
        tempDt.yy = gregYr + 3761;        /* next year */
        *endDST = day_on_or_before (SUN, hebrew2abs (tempDt) - 3L);
        DST_value = 60;
        break;
    case DST_NONE:
        *endDST = *beginDST = 0L;
        DST_value = 0;
        break;
    default:
        DST_value = 0;
        break;
    }
}

/*-------------------------------------------------------------------------*/

void 
set_DST( long beginDST, long endDST, long todayAbs, int *DST )
{
    if (beginDST > endDST)
    {
	/* Australia */
	if (todayAbs >= beginDST || todayAbs < endDST)
	{
	    *DST = DST_value;
	}
	else
	{
	    *DST = 0;
	}
    }
    else
    {
	if (beginDST <= todayAbs && todayAbs < endDST)
	    *DST = DST_value;
	else
	    *DST = 0;
    }
}

/*-------------------------------------------------------------------------*/

void print_candlelighting_times( int mask, int weekday, date_t todayGreg, int DST)
{
    /* offset of sunset to candlelighting in minutes */
    
    if (weekday == FRI || (mask & LIGHT_CANDLES))
    {
        double xsunrise, xsunset;
        int day_adj, status;
        
        PrintGregDate (todayGreg);
        
        status = suntime (&xsunrise, &xsunset, dayOfYear (todayGreg),
                        SUNRISE, SUNSET);
        if (status & NO_SUNSET)
            printf ("No sunset today.\n");
        else
        {
            char * time =timeadj ("", xsunset, light_offset + DST, &day_adj);
            printf ("%s: %s\n",
                    iso8859_8_sw ? "\344\343\354\367\372 \360\370\345\372" :
                    "Candle lighting",
                    time
                );
            free( time );
            /*              printf("%s %s %.1lf\n",timeadj ("",xsunrise,0,&day_adj),
                            timeadj ("",xsunset,0,&day_adj),
                            (xsunset-xsunrise) * 5.0 + 120.0);
            */
        }
    }
    
    /* offset of sunset to havdallah in minutes */
    if (weekday == SAT ||
        ((mask & YOM_TOV_ENDS) &&
        weekday != FRI))
    {
        double xsunrise, xsunset;
        int day_adj, status;
        
      
        PrintGregDate (todayGreg);
        
        status = suntime (&xsunrise, &xsunset, dayOfYear (todayGreg),
                          SUNRISE, SUNSET);
        if (status & NO_SUNSET)
            printf ("No sunset today.\n");
        else
        {
            char * time = timeadj ("",
                                   xsunset,
                                   havdalah_minutes + DST,
                                   &day_adj);
            printf ("%s (%d %s):%s\n",
                    iso8859_8_sw ? "\344\341\343\354\344" : "Havdalah",
                    havdalah_minutes,
                    iso8859_8_sw ? "\343\367\345\372" : "min",
                    time
                    );
            free( time );
        }
    }
}

/*-------------------------------------------------------------------------*/

void reset_Omer( int hYear )
{
    date_t d;
    d.dd = 16;
    d.mm = NISAN;
    d.yy = hYear;

    beginOmer = hebrew2abs (d);
    d.dd = 5;
    d.mm = SIVAN;
    endOmer = hebrew2abs (d);
}

/*-------------------------------------------------------------------------*/

void main_calendar( long todayAbs, long endAbs) /* the range of the desired printout */
{
    date_t todayGreg, todayHeb;
    long beginDST, endDST;
    holstorep_t holi_start,holip;         /* a list of holidays for today */
    year_t theYear;
    char *omerStr ;
    int omer, day_of_week, returnedMask, DST;
    int omer_today, sedra_today, candle_today, holidays_today;
    
    todayHeb = abs2hebrew (todayAbs);
    todayGreg = abs2greg (todayAbs);
    
    theYear = yearData (todayHeb.yy);
    set_DST_bounds (&beginDST, &endDST, todayGreg.yy);
    set_DST (beginDST, endDST, todayAbs, &DST);

    /*============== Main Year Loop ==============*/
    
    
    reset_Omer (todayHeb.yy);
    if (sedraAllWeek_sw || sedrot_sw)
        reset_sedra (todayHeb.yy);
    day_of_week = (int) (todayAbs % 7L);
    while (todayAbs <= endAbs)
    {
        /* get the holidays for today */
      returnedMask = getHebHolidays (todayHeb, &holip);
      
      sedra_today = sedraAllWeek_sw || (sedrot_sw && (day_of_week == SAT));
      omer_today = printOmer_sw &&
          (todayAbs >= beginOmer) &&
          (todayAbs <= endOmer);
      candle_today = candleLighting_sw &&
          (day_of_week >= FRI ||
           (returnedMask & LIGHT_CANDLES) ||
           (returnedMask & YOM_TOV_ENDS));
      holidays_today = holip &&
        (!noHolidays_sw || (returnedMask & USER_EVENT));
      
      if (printHebDates_sw ||
          (printSomeHebDates_sw && 
           (holidays_today || sedra_today || omer_today || candle_today)))
      {
          PrintGregDate (todayGreg);
          printf ("%d%s%s %s, %d\n", todayHeb.dd,       /* print the hebrew date */
                  iso8859_8_sw ? "" : numSuffix( todayHeb.dd ),
                  iso8859_8_sw ? "" : " of",
                  LANGUAGE2(hMonths[LEAP_YR_HEB( todayHeb.yy )][todayHeb.mm].name),
                  todayHeb.yy);
      }
      
      
      /* print the sedra, if desired */
      if (sedra_today)
      {
          const char *sedraStr = sedra( todayAbs );
          if (NULL != sedraStr)
          {
              PrintGregDate( todayGreg );
              printf( "%s %s\n",
                      iso8859_8_sw ? "\364\370\371\372" :
                      ashkenazis_sw ? "Parshas" : "Parashat",
                      sedraStr );
          }
      }
      
      /* print today's holidays */
      holi_start=holip;         /* store the head of the list for freeing */
      for (; holip; holip = holip->next)
      {
          if (!noHolidays_sw || (holip->typeMask & USER_EVENT))
          {
/*              char buf[200];
              puts(formatLine( todayGreg, todayHeb, holip->name, buf, 200));
*/
              PrintGregDate( todayGreg ); 
              puts( holip->name ); 
          }
      }
      
      /* Print the Omer */
      if (omer_today)
      {
          initStr (&omerStr, NM_LEN);
          omer = (int) (todayAbs - beginOmer + 1L);
          if (!tabs_sw)
          {
              strncat (omerStr, hc_itoa (omer), NM_LEN);
              strncat (omerStr, numSuffix (omer), NM_LEN);
              strncat (omerStr, " day of the Omer", NM_LEN);
          }
          else
          {
              strncat (omerStr, "Omer: ", NM_LEN);
              strncat (omerStr, hc_itoa (omer), NM_LEN);
          }
          PrintGregDate (todayGreg);
          printf ("%s\n", omerStr);
          free( omerStr );
      }
      
      /* Print CandleLighting times  */
      if (candle_today)
      {
          set_DST (beginDST, endDST, todayAbs, &DST);
          print_candlelighting_times (returnedMask,
                                      day_of_week, todayGreg, DST);
      }
      

      incHebGregDate (&todayHeb, &todayGreg, &todayAbs, &day_of_week, &theYear);
      
      if (1 == todayGreg.dd &&  /* happy new year! */
          JAN == todayGreg.mm)
      {
          set_DST_bounds (&beginDST, &endDST, todayGreg.yy);
      }
      
#     ifdef PLUG_LEAKS
      free_holidays(*holip);
#     endif
    }
}



